src/lib/formats/rpk.cpp

```diff
@@ -0,0 +1,459 @@
+		{
+			std::error_condition err = file->add_rom_socket(std::string(*id), *resource_node);
+			if (err)
+				return err;
+		}
+		else if (!strcmp(resource_node->get_name(), "ram"))
+		{
+			if (!m_supports_ram)
+				return error::UNSUPPORTED_RPK_FEATURE; // <ram> is not supported by this system
+			std::error_condition err = file->add_ram_socket(std::string(*id), *resource_node);
+			if (err)
+				return err;
+		}
+		else
+			return error::INVALID_LAYOUT; // resource node must be <rom> or <ram>
+	}
+
+	// and we're done!
+	result = std::move(file);
+	return std::error_condition();
+}
+
+
+/***************************************************************************
+	RPK FILE
+***************************************************************************/
+
+//-------------------------------------------------
+//  ctor
+//-------------------------------------------------
+
+util::rpk_file::rpk_file(archive_file::ptr &&zipfile, int pcb_type)
+	: m_zipfile(std::move(zipfile))
+	, m_pcb_type(pcb_type)
+{
+}
+
+
+
+//-------------------------------------------------
+//  dtor
+//-------------------------------------------------
+
+util::rpk_file::~rpk_file()
+{
+}
+
+
+//-------------------------------------------------
+//  add_rom_socket
+//-------------------------------------------------
+
+std::error_condition util::rpk_file::add_rom_socket(std::string &&id, const util::xml::data_node &rom_resource_node)
+{
+	// find the file attribute (required)
+	std::string const *const file = rom_resource_node.get_attribute_string_ptr("file");
+	if (!file)
+		return rpk_reader::error::INVALID_LAYOUT; // <rom> must have a 'file' attribute
+
+	// check for crc (optional)
+	std::optional<hash_collection> hashes = { };
+	std::string const *const crcstr = rom_resource_node.get_attribute_string_ptr("crc");
+	if (crcstr)
+	{
+		if (!hashes.has_value())
+			hashes.emplace();
+		hashes->add_from_string(hash_collection::HASH_CRC, *crcstr);
+	}
+
+	// check for sha1 (optional)
+	std::string const *const sha1 = rom_resource_node.get_attribute_string_ptr("sha1");
+	if (sha1)
+	{
+		if (!hashes.has_value())
+			hashes.emplace();
+		hashes->add_from_string(hash_collection::HASH_SHA1, *sha1);
+	}
+
+	// finally add the socket
+	m_sockets.emplace_back(*this, std::move(id), rpk_socket::socket_type::ROM, std::string(*file), std::move(hashes), std::nullopt);
+	return std::error_condition();
+}
+
+
+//-------------------------------------------------
+//  add_ram_socket
+//-------------------------------------------------
+
+std::error_condition util::rpk_file::add_ram_socket(std::string &&id, const util::xml::data_node &ram_resource_node)
+{
+	// find the length attribute
+	std::string const *const length_string = ram_resource_node.get_attribute_string_ptr("length");
+	if (!length_string)
+		return rpk_reader::error::MISSING_RAM_LENGTH;
+
+	// parse it
+	unsigned int length;
+	char suffix = '\0';
+	sscanf(length_string->c_str(), "%u%c", &length, &suffix);
```

Comment:
You should test the result of `sscanf` – it tells you how many of the format arguments were processed.